### PR TITLE
Try to recover from fatal JIT errors.

### PR DESCRIPTION
**Pull request by hvdijk**
# Overview

Try to recover from fatal JIT errors.

# Reason for change

On ARM, the following disabled UnitCL tests result in errors that take down the whole UnitCL process:

  Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_2
  Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_2
  Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_2
  Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_2
  Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_2

# Description of change

With this change, after those tests fail, UnitCL is able to continue running to cover the remaining tests.

Note that we do fail an assertion at global test environment tear-down. This happens because CrashRecoveryContext is implemented via setjmp/longjmp, which is not required to call destructors for local objects. Likely, this already resulted in memory leaks elsewhere, but here, we leak a reference-counted string, and the string pool asserts that no references remain in its destructor. This is not ideal, but I see no good way around this, crashing there is more useful for us than crashing right away, and in builds with assertions disabled, it will not start doing the wrong thing.

# Anything else we should know?

*If there's any other relevant information we should know that may help us in
understanding and verifying your patch, please include it here.*

# Checklist

* Read and follow the project [Code of Conduct](https://github.com/codeplaysoftware/oneapi-construction-kit/blob/main/CODE_OF_CONDUCT.md).
* Make sure the project builds successfully with your changes.
* Run relevant testing locally to avoid regressions.
* Run [clang-format-16](https://clang.llvm.org/docs/ClangFormat.html) (the most
  recent version available through `pip`) on all modified code.

### CODE REVIEW

File: modules/compiler/targets/host/source/kernel.cpp [hunk 1]
```diff
@@ -388,8 +388,62 @@ HostKernel::lookupOrCreateOptimizedKernel(std::array<size_t, 3> local_size) {
...
+      // over destructors of objects that do need to be destroyed. We do so
+      // anyway because the effect is less bad than crashing right away.
+      std::promise<uint64_t> promise;
+      llvm::Error err = llvm::Error::success();
+      llvm::cantFail(std::move(err));
+
+      auto &ES = target.orc_engine->getExecutionSession();
+      auto SO = makeJITDylibSearchOrder(
+          &*jd, llvm::orc::JITDylibLookupFlags::MatchAllSymbols);
+      auto Name = ES.intern(target.orc_engine->mangle(unique_name));
```
Comment (frasercrmck) [inline modules/compiler/targets/host/source/kernel.cpp:403]: I know we're not in a consistent place across the project, but this function seems to have settled on a naming scheme which `Name` isn't adhering to (in my opinion).

Obviously it's arbitrary but I think `CRC` is probably fine for its conciseness, but I think `name` and `names` and `notify_complete` might suit better? I realise we have `globalLock` above too. I'm not sure about `ES` or `SO`. Ugh..

**Review by frasercrmck — COMMENTED** [review]
Seems broadly fine to me.

I'd quite like to see which errors we're swapping in favour of which other errors, just so I'm prepared for what we're getting into. That might be quite hard to convey, though.

File: modules/compiler/targets/host/source/kernel.cpp (hunk 1 above)
Comment (hvdijk) [reply modules/compiler/targets/host/source/kernel.cpp:403]: Ah, right, this is basically copied and modified from LLVM code and I didn't adjust the LLVM naming of things. Will do.

**Comment by hvdijk** [comment]
> I'd quite like to see which errors we're swapping in favour of which other errors, just so I'm prepared for what we're getting into. That might be quite hard to convey, though.

I'll include the full output from a (filtered) `UnitCL` run with affected tests forcibly enabled, that should make it clearer than the description I have now.

**Comment by frasercrmck** [comment]
> > I'd quite like to see which errors we're swapping in favour of which other errors, just so I'm prepared for what we're getting into. That might be quite hard to convey, though.
> 
> I'll include the full output from a (filtered) `UnitCL` run with affected tests forcibly enabled, that should make it clearer than the description I have now.

That'd be great, yeah.

**Comment by hvdijk** [comment]
With the tests disabled with a comment saying CA-2882 re-enabled, without this PR, we get:
```
$ OCL_ICD_VENDORS=/dev/null OCL_ICD_FILENAMES=$PWD/lib/libCL.so qemu-arm -L /usr/arm-linux-gnueabihf bin/UnitCL --gtest_filter='Execution/HalfOperatorTest.Precision_0*'
Random numbers generated using std::mt19937 with seed 1386920080
Note: Google Test filter = Execution/HalfOperatorTest.Precision_0*
[==========] Running 30 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 30 tests from Execution/HalfOperatorTest
[ RUN      ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_0
[       OK ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_0 (1417 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_1
[       OK ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_1 (219 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_2
SplitVectorResult #0: t18: v4f16,v4f16,v4f16,ch = llvm.arm.neon.vld3<(load (s192) from %ir.vload_base.i.i.i, align 2, !noalias !32, addrspace 1)> t0, TargetConstant:i32<2984>, t15, Constant:i32<2>

LLVM ERROR: Do not know how to split the result of this operator!

qemu: uncaught target signal 6 (Aborted) - core dumped
Aborted
```
With this PR, we get:
```
$ OCL_ICD_VENDORS=/dev/null OCL_ICD_FILENAMES=$PWD/lib/libCL.so qemu-arm -L /usr/arm-linux-gnueabihf bin/UnitCL --gtest_filter='Execution/HalfOperatorTest.Precision_0*'
Random numbers generated using std::mt19937 with seed 1372472470
Note: Google Test filter = Execution/HalfOperatorTest.Precision_0*
[==========] Running 30 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 30 tests from Execution/HalfOperatorTest
[ RUN      ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_0
[       OK ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_0 (1413 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_1
[       OK ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_1 (222 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_2
SplitVectorResult #0: t18: v4f16,v4f16,v4f16,ch = llvm.arm.neon.vld3<(load (s192) from %ir.vload_base.i.i.i, align 2, !noalias !32, addrspace 1)> t0, TargetConstant:i32<2984>, t15, Constant:i32<2>

LLVM ERROR: Do not know how to split the result of this operator!

Failed to materialize symbols: { (__mux_host_2.dylib, { __mux_host_2 }) }
/home/harald/oneapi-construction-kit/modules/kts/include/kts/execution_shared.h:36: Failure
Could not enqueue N-D range (error: -44)
[  FAILED  ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_2, where GetParam() = (0, 3) (213 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_3
[       OK ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_3 (222 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_4
[       OK ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_4 (421 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_5
[       OK ] Execution/HalfOperatorTest.Precision_03_Half_Add/OpenCLC_5 (444 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_0
[       OK ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_0 (161 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_1
[       OK ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_1 (183 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_2
SplitVectorResult #0: t18: v4f16,v4f16,v4f16,ch = llvm.arm.neon.vld3<(load (s192) from %ir.vload_base.i.i.i, align 2, !noalias !32, addrspace 1)> t0, TargetConstant:i32<2984>, t15, Constant:i32<2>

LLVM ERROR: Do not know how to split the result of this operator!

Failed to materialize symbols: { (__mux_host_8.dylib, { __mux_host_8 }) }
/home/harald/oneapi-construction-kit/modules/kts/include/kts/execution_shared.h:36: Failure
Could not enqueue N-D range (error: -44)
[  FAILED  ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_2, where GetParam() = (0, 3) (156 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_3
[       OK ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_3 (207 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_4
[       OK ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_4 (389 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_5
[       OK ] Execution/HalfOperatorTest.Precision_04_Half_Sub/OpenCLC_5 (438 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_0
[       OK ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_0 (161 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_1
[       OK ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_1 (184 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_2
SplitVectorResult #0: t18: v4f16,v4f16,v4f16,ch = llvm.arm.neon.vld3<(load (s192) from %ir.vload_base.i.i.i, align 2, !noalias !32, addrspace 1)> t0, TargetConstant:i32<2984>, t15, Constant:i32<2>

LLVM ERROR: Do not know how to split the result of this operator!

Failed to materialize symbols: { (__mux_host_14.dylib, { __mux_host_14 }) }
/home/harald/oneapi-construction-kit/modules/kts/include/kts/execution_shared.h:36: Failure
Could not enqueue N-D range (error: -44)
[  FAILED  ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_2, where GetParam() = (0, 3) (155 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_3
[       OK ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_3 (209 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_4
[       OK ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_4 (387 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_5
[       OK ] Execution/HalfOperatorTest.Precision_05_Half_Mul/OpenCLC_5 (439 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_0
[       OK ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_0 (161 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_1
[       OK ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_1 (181 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_2
SplitVectorResult #0: t18: v4f16,v4f16,v4f16,ch = llvm.arm.neon.vld3<(load (s192) from %ir.vload_base.i.i.i, align 2, !noalias !32, addrspace 1)> t0, TargetConstant:i32<2984>, t15, Constant:i32<2>

LLVM ERROR: Do not know how to split the result of this operator!

Failed to materialize symbols: { (__mux_host_20.dylib, { __mux_host_20 }) }
/home/harald/oneapi-construction-kit/modules/kts/include/kts/execution_shared.h:36: Failure
Could not enqueue N-D range (error: -44)
[  FAILED  ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_2, where GetParam() = (0, 3) (155 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_3
[       OK ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_3 (205 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_4
[       OK ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_4 (365 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_5
[       OK ] Execution/HalfOperatorTest.Precision_06_Half_Div/OpenCLC_5 (417 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_0
[       OK ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_0 (153 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_1
[       OK ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_1 (231 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_2
SplitVectorResult #0: t18: v4f16,v4f16,v4f16,ch = llvm.arm.neon.vld3<(load (s192) from %ir.vload_base.i.i.i, align 2, !noalias !32, addrspace 1)> t0, TargetConstant:i32<2984>, t15, Constant:i32<2>

LLVM ERROR: Do not know how to split the result of this operator!

Failed to materialize symbols: { (__mux_host_26.dylib, { __mux_host_26 }) }
/home/harald/oneapi-construction-kit/modules/kts/include/kts/execution_shared.h:36: Failure
Could not enqueue N-D range (error: -44)
[  FAILED  ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_2, where GetParam() = (0, 3) (145 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_3
[       OK ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_3 (238 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_4
[       OK ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_4 (207 ms)
[ RUN      ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_5
[       OK ] Execution/HalfOperatorTest.Precision_07_Half_Recip/OpenCLC_5 (327 ms)
[----------] 30 tests from Execution/HalfOperatorTest (8794 ms total)

[----------] Global test environment tear-down
UnitCL: /home/harald/llvm-project/llvm/include/llvm/ExecutionEngine/Orc/SymbolStringPool.h:240: llvm::orc::SymbolStringPool::~SymbolStringPool(): Assertion `Pool.empty() && "Dangling references at pool destruction time"' failed.
qemu: uncaught target signal 6 (Aborted) - core dumped
Aborted
```

File: modules/compiler/targets/host/source/kernel.cpp (hunk 1 above)
Comment (hvdijk) [reply modules/compiler/targets/host/source/kernel.cpp]: Changed everything to start lowercase to mostly fit the existing style. Given that this is a function named `lookupOrCreateOptimizedKernel`, I think the style for multi-word names needs to be camel case even if not everything else in the function follows that same style, does that sound okay to you?

**Review by frasercrmck — APPROVED** [review]
LGTM. I agree it's most likely an improvement.

Hopefully we can resolve the whole issue "cleanly" at some point too, even if that requires work on LLVM itself.

**Pull request merged**